File: discipline_kt/app/src/main/java/com/example/app/chronic/Duration.kt
package com.example.app

import com.example.app.*

@JvmInline
public value class Duration private constructor(val milliseconds: Long) {
  companion object {
    const val MILLISECONDS_PER_SECOND = 1000L
    const val MILLISECONDS_PER_MINUTE = MILLISECONDS_PER_SECOND * 60
    const val MILLISECONDS_PER_HOUR = MILLISECONDS_PER_MINUTE * 60
    const val MILLISECONDS_PER_DAY = MILLISECONDS_PER_HOUR * 24
    const val MILLISECONDS_PER_WEEK = MILLISECONDS_PER_DAY * 7

    const val MAXIMUM_MILLISECONDS = Long.MAX_VALUE
    const val MAXIMUM_SECONDS = MAXIMUM_MILLISECONDS / MILLISECONDS_PER_SECOND
    const val MAXIMUM_MINUTES = MAXIMUM_SECONDS / 60
    const val MAXIMUM_HOURS = MAXIMUM_MINUTES / 60
    const val MAXIMUM_DAYS = MAXIMUM_HOURS / 24
    const val MAXIMUM_WEEKS = MAXIMUM_DAYS / 7

    val DAY = Duration(MILLISECONDS_PER_DAY)

    /**
     * Creates a Duration from milliseconds
     */
    fun fromMilliseconds(
      milliseconds: Long,
    ): Tried<Duration, TextualError> {
      if (milliseconds < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'milliseconds'", milliseconds)
        )
      }
      
      if (milliseconds > MAXIMUM_MILLISECONDS) {
        return Tried.failure(
          TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is greater than maximum value")
          .addLongAttachment("Argument 'milliseconds'", milliseconds)
          .addLongAttachment("Maximum value", MAXIMUM_MILLISECONDS)
        )
      }

      return Tried.success(Duration(milliseconds))
    }

    fun fromMillisecondsOrThrow(milliseconds: Long): Duration {
      if (milliseconds < 0) {
        throw TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'milliseconds'", milliseconds)
      }

      if (milliseconds > MAXIMUM_MILLISECONDS) {
        throw TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is greater than maximum value")
          .addLongAttachment("Argument 'milliseconds'", milliseconds)
          .addLongAttachment("Maximum value", MAXIMUM_MILLISECONDS)
      }

      return Duration(milliseconds)
    }

    /**
     * Creates a Duration from seconds
     */
    fun fromSeconds(
      seconds: Long,
    ): Tried<Duration, TextualError> {
      if (seconds < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'seconds'", seconds)
        )
      }
      
      if (seconds > MAXIMUM_SECONDS) {
        return Tried.failure(
          TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is greater than maximum value")
          .addLongAttachment("Argument 'seconds'", seconds)
          .addLongAttachment("Maximum value", MAXIMUM_SECONDS)
        )
      }

      return Tried.success(Duration(seconds * MILLISECONDS_PER_SECOND))
    }

    fun fromSecondsOrThrow(
      seconds: Long,
    ): Duration {
      if (seconds < 0) {
        throw TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'seconds'", seconds)
      }

      if (seconds > MAXIMUM_SECONDS) {
        throw TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is greater than maximum value")
          .addLongAttachment("Argument 'seconds'", seconds)
          .addLongAttachment("Maximum value", MAXIMUM_SECONDS)
      }

      return Duration(seconds * MILLISECONDS_PER_SECOND)
    }

    /**
     * Creates a Duration from minutes
     */
    fun fromMinutes(
      minutes: Long,
    ): Tried<Duration, TextualError> {
      if (minutes < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'minutes'", minutes)
        )
      }
      
      if (minutes > MAXIMUM_MINUTES) {
        return Tried.failure(
          TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is greater than maximum value")
          .addLongAttachment("Argument 'minutes'", minutes)
          .addLongAttachment("Maximum value", MAXIMUM_MINUTES)
        )
      }

      return Tried.success(Duration(minutes * MILLISECONDS_PER_MINUTE))
    }

    fun fromMinutesOrThrow(
      minutes: Long,
    ): Duration {
      if (minutes < 0) {
        throw TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'minutes'", minutes)
      }

      if (minutes > MAXIMUM_MINUTES) {
        throw TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is greater than maximum value")
          .addLongAttachment("Argument 'minutes'", minutes)
          .addLongAttachment("Maximum value", MAXIMUM_MINUTES)
      }

      return Duration(minutes * MILLISECONDS_PER_MINUTE)
    }

    /**
     * Creates a Duration from hours
     */
    fun fromHours(
      hours: Long,
    ): Tried<Duration, TextualError> {
      if (hours < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'hours'", hours)
        )
      }
      
      if (hours > MAXIMUM_HOURS) {
        return Tried.failure(
          TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is greater than maximum value")
          .addLongAttachment("Argument 'hours'", hours)
          .addLongAttachment("Maximum value", MAXIMUM_HOURS)
        )
      }

      return Tried.success(Duration(hours * MILLISECONDS_PER_HOUR))
    }

    fun fromHoursOrThrow(
      hours: Long,
    ): Duration {
      if (hours < 0) {
        throw TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'hours'", hours)
      }

      if (hours > MAXIMUM_HOURS) {
        throw TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is greater than maximum value")
          .addLongAttachment("Argument 'hours'", hours)
          .addLongAttachment("Maximum value", MAXIMUM_HOURS)
      }

      return Duration(hours * MILLISECONDS_PER_HOUR)
    }

    /**
     * Creates a Duration from days
     */
    fun fromDays(
      days: Long,
    ): Tried<Duration, TextualError> {
      if (days < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'days'", days)
        )
      }
      
      if (days > MAXIMUM_DAYS) {
        return Tried.failure(
          TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is greater than maximum value")
          .addLongAttachment("Argument 'days'", days)
          .addLongAttachment("Maximum value", MAXIMUM_DAYS)
        )
      }

      return Tried.success(Duration(days * MILLISECONDS_PER_DAY))
    }

    fun fromDaysOrThrow(
      days: Long,
    ): Duration {
      if (days < 0) {
        throw TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'days'", days)
      }

      if (days > MAXIMUM_DAYS) {
        throw TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is greater than maximum value")
          .addLongAttachment("Argument 'days'", days)
          .addLongAttachment("Maximum value", MAXIMUM_DAYS)
      }

      return Duration(days * MILLISECONDS_PER_DAY)
    }

    /**
     * Creates a Duration from weeks
     */
    fun fromWeeks(
      weeks: Long,
    ): Tried<Duration, TextualError> {
      if (weeks < 0) {
        return Tried.failure(
          TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'weeks'", weeks)
        )
      }
      
      if (weeks > MAXIMUM_WEEKS) {
        return Tried.failure(
          TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is greater than maximum value")
          .addLongAttachment("Argument 'weeks'", weeks)
          .addLongAttachment("Maximum value", MAXIMUM_WEEKS)
        )
      }

      return Tried.success(Duration(weeks * MILLISECONDS_PER_WEEK))
    }

    fun fromWeeksOrThrow(
      weeks: Long,
    ): Duration {
      if (weeks < 0) {
        throw TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is negative: This Duration type only supports representing positive durations")
          .addLongAttachment("Argument 'weeks'", weeks)
      }

      if (weeks > MAXIMUM_WEEKS) {
        throw TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is greater than maximum value")
          .addLongAttachment("Argument 'weeks'", weeks)
          .addLongAttachment("Maximum value", MAXIMUM_WEEKS)
      }

      return Duration(weeks * MILLISECONDS_PER_WEEK)
    }

    /**
     * Returns a zero duration
     */
    fun zero(): Duration {
      return Duration(0)
    }
  }

  fun toTotalMilliseconds(): Long {
    return milliseconds
  }

  fun toTotalSeconds(): Long {
    return milliseconds / MILLISECONDS_PER_SECOND
  }

  fun toTotalMinutes(): Long {
    return this.milliseconds / MILLISECONDS_PER_MINUTE
  }

  fun toTotalHours(): Long {
    return this.milliseconds / MILLISECONDS_PER_HOUR
  }

  fun toTotalDays(): Long {
    return this.milliseconds / MILLISECONDS_PER_DAY
  }

  fun toTotalWeeks(): Long {
    return this.milliseconds / MILLISECONDS_PER_WEEK
  }


  fun isZero(): Boolean {
    return milliseconds == 0L
  }

  fun saturatingSub(rhs: Duration): Duration {
    return if (milliseconds > rhs.milliseconds) {
      Duration(milliseconds - rhs.milliseconds)
    } else {
      zero()
    }
  }

  fun saturatingAdd(rhs: Duration): Duration {
    return if (milliseconds >= MAXIMUM_MILLISECONDS - rhs.milliseconds) {
      Duration(MAXIMUM_MILLISECONDS)
    } else {
      Duration(milliseconds + rhs.milliseconds)
    }
  }

  /**
   * Returns true if this duration is equal to another
   */
  fun isEqualTo(rhs: Duration): Boolean {
    return milliseconds == rhs.milliseconds
  }

  /**
   * Returns true if this duration is longer than another
   */
  fun isLongerThan(rhs: Duration): Boolean {
    return milliseconds > rhs.milliseconds
  }

  /**
   * Returns true if this duration is longer than or equal to another
   */
  fun isLongerThanOrEqualTo(rhs: Duration): Boolean {
    return milliseconds >= rhs.milliseconds
  }

  /**
   * Returns true if this duration is shorter than another
   */
  fun isShorterThan(rhs: Duration): Boolean {
    return milliseconds < rhs.milliseconds
  }

  /**
   * Returns true if this duration is shorter than or equal to another
   */
  fun isShorterThanOrEqualTo(rhs: Duration): Boolean {
    return milliseconds <= rhs.milliseconds
  }

  /**
   * Returns a human-readable string representation (e.g., "5d 3h 2m")
   */
  override fun toString(): String {
    var remainingMs = milliseconds
    val parts = mutableListOf<String>()
    
    val days = remainingMs / MILLISECONDS_PER_DAY
    remainingMs %= MILLISECONDS_PER_DAY
    if (days > 0) parts.add("${days}d")
    
    val hours = remainingMs / MILLISECONDS_PER_HOUR
    remainingMs %= MILLISECONDS_PER_HOUR
    if (hours > 0) parts.add("${hours}h")
    
    val minutes = remainingMs / MILLISECONDS_PER_MINUTE
    remainingMs %= MILLISECONDS_PER_MINUTE
    if (minutes > 0) parts.add("${minutes}m")
    
    val seconds = remainingMs / MILLISECONDS_PER_SECOND
    if (seconds > 0 || parts.isEmpty()) parts.add("${seconds}s")
    
    return parts.joinToString(" ")
  }

  /**
   * Alternative string representation with full units (e.g., "5 Days 3 Hours 2 Minutes")
   */
  fun toLongString(): String {
    var remainingMs = milliseconds
    val parts = mutableListOf<String>()
    
    val days = remainingMs / MILLISECONDS_PER_DAY
    remainingMs %= MILLISECONDS_PER_DAY
    if (days > 0) parts.add("${days} D")
    
    val hours = remainingMs / MILLISECONDS_PER_HOUR
    remainingMs %= MILLISECONDS_PER_HOUR
    if (hours > 0) parts.add("${hours} H")
    
    val minutes = remainingMs / MILLISECONDS_PER_MINUTE
    remainingMs %= MILLISECONDS_PER_MINUTE
    if (minutes > 0) parts.add("${minutes} M")
    
    return parts.joinToString(" ")
  }

  fun min(rhs: Duration): Duration {
    return if (this.isShorterThan(rhs)) {
      this 
    } else {
      rhs
    }
  }

  fun max(rhs: Duration): Duration {
    return if (this.isLongerThan(rhs)) {
      this
    } else {
      rhs
    }
  }

  // operator fun plus(other: Duration): Duration {
  //   return saturatingAdd(other)
  // }
  // operator fun minus(other: Duration): Duration = saturatingSub(other)
  // operator fun compareTo(other: Duration): Int = milliseconds.compareTo(other.milliseconds)
}

File: discipline_ts/chronic/duration.ts
import { FAILURE, FailureCode, TextualError, Tried, Unique } from "../x.ts";

const BRAND = Symbol();

export type Duration = Unique<typeof BRAND, "Duration", number>;

const construct = (milliseconds: number): Duration => {
  return milliseconds as Duration;
};

const MAXIMUM_MILLISECONDS = Number.MAX_SAFE_INTEGER;
const MAXIMUM_SECONDS = Math.floor(MAXIMUM_MILLISECONDS / 1000);
const MAXIMUM_MINUTES = Math.floor(MAXIMUM_SECONDS / 60);
const MAXIMUM_HOURS = Math.floor(MAXIMUM_MINUTES / 60);
const MAXIMUM_DAYS = Math.floor(MAXIMUM_HOURS / 24);
const MAXIMUM_WEEKS = Math.floor(MAXIMUM_DAYS / 7);

const MILLISECONDS_PER_SECOND = 1000;
const MILLISECONDS_PER_MINUTE = MILLISECONDS_PER_SECOND * 60;
const MILLISECONDS_PER_HOUR = MILLISECONDS_PER_MINUTE * 60;
const MILLISECONDS_PER_DAY = MILLISECONDS_PER_HOUR * 24;
const MILLISECONDS_PER_WEEK = MILLISECONDS_PER_DAY * 7;

const DAY = construct(MILLISECONDS_PER_DAY);

const fromMilliseconds = (milliseconds: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(milliseconds)) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    return Tried.Failure(it);
  }
  
  if (milliseconds < 0) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    return Tried.Failure(it);
  }
  
  if (milliseconds > MAXIMUM_MILLISECONDS) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_MILLISECONDS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(milliseconds));
};

const fromMillisecondsOrErrorCode = (
  milliseconds: number,
  textualError: TextualError,
): Duration | FailureCode => {
  if (!Number.isInteger(milliseconds)) {
    TextualError.changeContext(textualError, "Creating a Duration from milliseconds");
    TextualError.addMessage(textualError, "Argument 'milliseconds' is not an integer")
    TextualError.addNumberAttachment(textualError, "Argument 'milliseconds'", milliseconds);
    return FAILURE;
  }
  
  if (milliseconds < 0) {
    TextualError.changeContext(textualError, "Creating a Duration from milliseconds");
    TextualError.addMessage(textualError, "Argument 'milliseconds' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(textualError, "Argument 'milliseconds'", milliseconds);
    return FAILURE;
  }
  
  if (milliseconds > MAXIMUM_MILLISECONDS) {
    TextualError.changeContext(textualError, "Creating a Duration from milliseconds");
    TextualError.addMessage(textualError, "Argument 'milliseconds' is greater than maximum value")
    TextualError.addNumberAttachment(textualError, "Argument 'milliseconds'", milliseconds);
    TextualError.addNumberAttachment(textualError, "Maximum value",  MAXIMUM_MILLISECONDS);
    return FAILURE;
  }

  return construct(milliseconds);
};

const fromSeconds = (seconds: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(seconds)) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    return Tried.Failure(it);
  }
  
  if (seconds < 0) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    return Tried.Failure(it);
  }
  
  if (seconds > MAXIMUM_SECONDS) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_SECONDS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(seconds * MILLISECONDS_PER_SECOND));
};

const fromMinutes = (minutes: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(minutes)) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    return Tried.Failure(it);
  }
  
  if (minutes < 0) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    return Tried.Failure(it);
  }
  
  if (minutes > MAXIMUM_MINUTES) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_MINUTES);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(minutes * MILLISECONDS_PER_MINUTE));
};

const fromHours = (hours: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(hours)) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    return Tried.Failure(it);
  }
  
  if (hours < 0) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    return Tried.Failure(it);
  }
  
  if (hours > MAXIMUM_HOURS) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_HOURS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(hours * MILLISECONDS_PER_HOUR));
};

const fromDays = (days: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(days)) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    return Tried.Failure(it);
  }
  
  if (days < 0) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    return Tried.Failure(it);
  }
  
  if (days > MAXIMUM_DAYS) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_DAYS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(days * MILLISECONDS_PER_DAY));
};

const fromWeeks = (weeks: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(weeks)) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    return Tried.Failure(it);
  }
  
  if (weeks < 0) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    return Tried.Failure(it);
  }
  
  if (weeks > MAXIMUM_WEEKS) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_WEEKS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(weeks * MILLISECONDS_PER_WEEK));
};

const fromMillisecondsOrThrow = (millseconds: number): Duration => {
  const it = fromMilliseconds(millseconds);
  if (Tried.isSuccess(it)) {
    return Tried.value(it);
  } else {
    throw new Error(TextualError.prettyPrint(Tried.error(it)));
  }
};

const zero = (): Duration => {
  return construct(0);
};

const isZero = (it: Duration): boolean => {
  return toTotalMilliseconds(it) === 0;
};

const toTotalMilliseconds = (it: Duration): number => {
  return it;
};

const toTotaMinutes = (it: Duration): number => {
  return Math.floor(toTotalMilliseconds(it) / MILLISECONDS_PER_MINUTE);
};

const minusOrZero = (it: Duration, rhs: Duration): Duration => {
  if (toTotalMilliseconds(it) > toTotalMilliseconds(rhs)) {
    return construct(toTotalMilliseconds(it) - toTotalMilliseconds(rhs));
  } else {
    return zero();
  }
};

const plusOrMax = (it: Duration, rhs: Duration): Duration => {
  const result = toTotalMilliseconds(it) + toTotalMilliseconds(rhs);
  if (result <= MAXIMUM_MILLISECONDS) {
    return construct(result);
  } else {
    return construct(MAXIMUM_MILLISECONDS);
  }
};

const isEqualTo = (lhs: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(lhs) === toTotalMilliseconds(rhs);
};

const isLongerThan = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) > toTotalMilliseconds(rhs);
};

const isLongerThanOrEqualTo = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) >= toTotalMilliseconds(rhs);
};

const isShorterThan = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) < toTotalMilliseconds(rhs);
};

const isShorterThanOrEqualTo = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) <= toTotalMilliseconds(rhs);
};

const min = (lhs: Duration, rhs: Duration): Duration => {
  return isShorterThan(lhs, rhs) ? lhs : rhs;
};

const max = (lhs: Duration, rhs: Duration): Duration => {
  return isLongerThan(lhs, rhs) ? lhs : rhs;
};

const toString = (it: Duration): string => {
  const parts: string[] = [];

  let totalMilliseconds = toTotalMilliseconds(it);
  
  const totalDays = Math.floor(totalMilliseconds / MILLISECONDS_PER_DAY);
  totalMilliseconds %= MILLISECONDS_PER_DAY;
  parts.push(`${totalDays} D`);
  
  const totalHours = Math.floor(totalMilliseconds / MILLISECONDS_PER_HOUR);
  totalMilliseconds %= MILLISECONDS_PER_HOUR;
  parts.push(`${totalHours} H`);
  
  const totalMinutes = Math.floor(totalMilliseconds / MILLISECONDS_PER_MINUTE);
  totalMilliseconds %= MILLISECONDS_PER_MINUTE;
  parts.push(`${totalMinutes} M`);
  
  return parts.join(' ');
};

const toString2 = (it: Duration): string => {
  const milliseconds = toTotalMilliseconds(it);

  if (milliseconds === 0) {
    return '0s';
  }

  const totalSeconds = Math.floor(milliseconds / 1000);
  const totalMinutes = Math.floor(totalSeconds / 60);
  const totalHours = Math.floor(totalMinutes / 60);
  const totalDays = Math.floor(totalHours / 24);

  const seconds = totalSeconds % 60;
  const minutes = totalMinutes % 60;
  const hours = totalHours % 24;
  const days = totalDays;

  const parts: string[] = [];

  if (days > 0) {
    parts.push(`${days}d`);
  }
  if (hours > 0) {
    parts.push(`${hours}h`);
  }
  if (minutes > 0) {
    parts.push(`${minutes}m`);
  }
  if (seconds > 0 || parts.length === 0) {
    parts.push(`${seconds}s`);
  }

  return parts.join(' ');
};

export const Duration = {
  MAXIMUM_MILLISECONDS,
  MAXIMUM_SECONDS,
  MAXIMUM_MINUTES,
  MAXIMUM_HOURS,
  MILLISECONDS_PER_SECOND,
  MILLISECONDS_PER_MINUTE,
  MILLISECONDS_PER_HOUR,
  MILLISECONDS_PER_DAY,
  MILLISECONDS_PER_WEEK,
  DAY,
  fromMilliseconds,
  fromMillisecondsOrErrorCode,
  fromMinutes,
  fromSeconds,
  fromHours,
  fromDays,
  fromWeeks,
  zero,
  toTotalMilliseconds,
  toTotaMinutes,
  isZero,
  saturatingSub: minusOrZero,
  saturatingAdd: plusOrMax,
  isEqualTo,
  isLongerThan,
  isLongerThanOrEqualTo,
  isShorterThan,
  isShorterThanOrEqualTo,
  min,
  max,
  toString,
  toString2,
  fromMillisecondsOrThrow,
};